eoflow/input/eopatch.py

```diff
@@ -0,0 +1,141 @@
+
+        return out_data
+
+    dataset = dataset.map(_read_patch)
+    return dataset
+
+
+class EOPatchInputExample(BaseInput):
+    """ An example input method. Shows reading EOPatches, subpatch extraction, data augmentation, caching, batching, etc. """
+
+    class _Schema(Schema):
+        data_dir = fields.String(description="The directory containing EOPatches.", required=True)
+
+        input_feature_type = fields.String(description="Feature type of the input feature.", required=True, validate=OneOf(_valid_types))
+        input_feature_name = fields.String(description="Name of the input feature.", required=True)
+        input_feature_axis = fields.List(fields.Int, description="Height and width axis for the input features", required=True, example=[1,2])
+        input_feature_shape = fields.List(fields.Int, description="Shape of the input feature. Use -1 for unknown dimesnions.",
+                                          required=True, example=[-1, 100, 100, 3])
+
+        labels_feature_type = fields.String(description="Feature type of the labels feature.", required=True, validate=OneOf(_valid_types))
+        labels_feature_name = fields.String(description="Name of the labels feature.", required=True)
+        labels_feature_axis = fields.List(fields.Int, description="Height and width axis for the labels", required=True, example=[1,2])
+        labels_feature_shape = fields.List(fields.Int, description="Shape of the labels feature. Use -1 for unknown dimesnions.",
+                                           required=True, example=[-1, 100, 100, 3])
+
+        patch_size = fields.List(fields.Int, description="Width and height of extracted patches.", required=True, example=[1,2])
+
+        interleave_size = fields.Int(description="Number of eopatches to interleave the subpatches from.", required=True, example=5)
+        batch_size = fields.Int(description="Number of examples in a batch.", required=True, example=20)
+        num_classes = fields.Int(description="Number of classes. Used for one-hot encoding.", required=True, example=2)
+
+        cache_file = fields.String(
+            missing=None, description="A path to the file where the dataset will be cached. No caching if not provided.", example='/tmp/data')
+        num_subpatches = fields.Int(required=True, description="Number of subpatches extracted by random sampling.", example=5)
+
+    def _parse_shape(self, shape):
+        shape = [None if s<0 else s for s in shape]
+        return shape
+
+    def get_dataset(self):
+        cfg = self.config
+
+        # Create a tf.data.Dataset from EOPatches
+        features_data = [
+            (cfg.input_feature_type, cfg.input_feature_name, 'features', np.float32, self._parse_shape(cfg.input_feature_shape)),
+            (cfg.labels_feature_type, cfg.labels_feature_name, 'labels', np.int64, self._parse_shape(cfg.labels_feature_shape))
+        ]
+        dataset = eopatch_dataset(self.config.data_dir, features_data, fill_na=-2)
```

Comment:
This could be made a general class to interface EOPatches, by exposing some other parameter such as `fill_na`, `random_sampling` and the params required by the grid splitting